keps/sig-storage/20200427-csi-driver-skip-permission.md

```diff
@@ -0,0 +1,155 @@
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Graduation Criteria](#graduation-criteria)
+  - [Test Plan](#test-plan)
+  - [Monitoring](#monitoring)
+- [Implementation History](#implementation-history)
+- [Drawbacks [optional]](#drawbacks-optional)
+- [Alternatives [optional]](#alternatives-optional)
+- [Infrastructure Needed [optional]](#infrastructure-needed-optional)
+<!-- /toc -->
+
+## Summary
+
+Currently before bind mounting a CSI volume for a Pod, we use imprecise heuristics,
+such as presence of fsType on the PVC to determine if the volume supports fsGroup based
+permission change. These heuristics are known to be fragile, and cause problems with different
+storage types.
+
+To solve this issue we will add a new field called `CSIDriver.Spec.SupportsFSGroup` 
+that allows the driver to define if it supports volume ownership modifications via
+fsGroup.
+
+## Motivation
+
+When a volume is mounted on the node, we modify volume ownership 
+before bind mounting the volume inside container; however, if the volume
+does not support these operations (such as NFS), then the change is still attempted. 
+This results in errors being reported to the user if the volume doesn't 
+support these operations.
+
+### Goals
+
+ - Allow CSIDrivers to opt-in to volume ownership changes.
+
+### Non-Goals
+
+ - This does not attempt to update existing CSIDrivers themselves.
+
+## Proposal
+
+We propose that the `CSIDriver` type include a field that defines if the volume 
+provided by the driver supports changing volume ownership. This will be enabled
+with a new feature gate, `CSIVolumeSupportFSGroup`.   
+
+### Implementation Details/Notes/Constraints [optional]
+
+Currently Volume permission and ownership change is done automatically for every volume. 
+As part of this proposal we will change the algorithm that modifies volume ownership and permissions
+for CSIDrivers to check the new field, and skip volume ownership modifications if it is found to be
+`false`.
+
+When defining a `CSIDriver`, we propose that `CSIDriver.Spec` be expanded to include a new field entitled 
+`SupportsFSGroup` which can have following possible values:
+
+ - `nil` --> Current behavior. Attempt to modify the volume ownership and permissions to the defined `fsGroup` when the volume is attached.
+ - `false` --> New behavior. Attach the volume without attempting to modify volume ownership or permissions.
+ - `true` --> New behavior. Always attempt to apply the defined fsGroup to modify volume ownership and permissions.
+
+```go
+type CSIDriverFSGroupChangePolicy string
+
+type CSIDriverSpec struct {
+    // SupportsFSGroup ← new field
+    // Defines if the underlying volume supports changing ownership and 
+    // permission of the volume before being attached. 
```

Comment:
"attached" or "mounted"?

Reply:
You're correct that this occurs in the csi_mounter code. I've updated this to be `mounted` instead of `attached.